### PR TITLE
[CALCITE-4102] Some improvements to aggregate related operations

### PR DESCRIPTION
**Pull request by liyafan82**
1. Improve the use of data structures
2. Extract common methods
3. Avoid expensive operations

### CODE REVIEW

**Comment by zinking** [comment]
looks all equivalent rewrites for the refactoring, 
any explanation for the improvements

File: core/src/main/java/org/apache/calcite/util/ImmutableBitSet.java [hunk 1]
```diff
@@ -901,6 +902,19 @@ public ImmutableBitSet shift(int offset) {
     return builder.build();
   }
 
+  /**
+   * Checks if all bit sets contain a particular bit.
+   */
+  public static boolean allContain(Collection<ImmutableBitSet> bitSets,
+                                   int bit) {
```
Comment (yanlin-Lynn) [inline core/src/main/java/org/apache/calcite/util/ImmutableBitSet.java:909]: we do not format like this

Comment (liyafan82) [reply core/src/main/java/org/apache/calcite/util/ImmutableBitSet.java]: Revised. Thank you for your kind reminder.

**Comment by liyafan82** [comment]
> looks all equivalent rewrites for the refactoring,
> any explanation for the improvements

Sure. Thanks a lot for your feedback. Below are some explanations:

1. We extract a common method for ImmutableBitSet, which is repeated multiple times in the code base. We also find it useful in our code base. This makes the code simpler and easier to maintain. 
2. In Aggregate#getRollup, we replace the LinkedHashSet with an ArrayList, because we can make sure that the newly added element must be distinct (according to the definition of a rollup). This makes adding data faster (LinkedHashSet takes O(logn) time, whereas ArrayList takes O(1)).
3. When checking if a group set is a roll up, we replace the except operation with cardinality minus, which is much faster. We can make this optimization, because when a set contains another, the cardinality of the difference is the difference of the cardiality.
4. In Util#firstDuplicate method, we replace the HashMap with a HashSet, because the value in the HashMap is not really used. This makes the logic clearer.

**Comment by yanlin-Lynn** [comment]
make sense.
Can you update the commit log to describe the content of the PR shortly,
and add your name in the end.

See the guide in https://calcite.apache.org/develop/#contributing

**Comment by liyafan82** [comment]
> make sense.
> Can you update the commit log to describe the content of the PR shortly,
> and add your name in the end.
> 
> See the guide in https://calcite.apache.org/develop/#contributing

@yanlin-Lynn Thanks a lot for your kind reminder. 
I have updated the commit message. Please check it.

**Comment by julianhyde** [comment]
FWIW, I intentionally used `HashMap` rather than `HashSet`. `HashSet` uses a `HashMap` under the covers. I figured I'd save an object allocation. Yours is a bit clearer, a bit less efficient.

`ret` is not a great name for a variable; also declare it as `List<Integer>` not `ArrayList<Integer>`.

**Comment by liyafan82** [comment]
> FWIW, I intentionally used `HashMap` rather than `HashSet`. `HashSet` uses a `HashMap` under the covers. I figured I'd save an object allocation. Yours is a bit clearer, a bit less efficient.

Thanks a lot for your feedback. It is clever to `HashMap` to avoid object allocation.
Fortunately, the value object for `HashSet` is declared static, so it is allocated once, and shared by all `HashSet` objects. 

> `ret` is not a great name for a variable; also declare it as `List<Integer>` not `ArrayList<Integer>`.

Good suggestion. I have renamed `ret` to `rollUpBits`.

File: core/src/main/java/org/apache/calcite/rel/core/Aggregate.java [hunk 2]
```diff
@@ -548,19 +538,22 @@ public static boolean isRollup(ImmutableBitSet groupSet,
...
+      final List<Integer> rollUpBits = new ArrayList<>(groupSets.size() - 1);
       ImmutableBitSet g = null;
       for (ImmutableBitSet bitSet : groupSets) {
         if (g == null) {
           // First item must equal groupSet
         } else {
           // Each subsequent items must be a subset with one fewer bit than the
           // previous item
-          set.addAll(g.except(bitSet).toList());
+          ImmutableBitSet diff = g.except(bitSet);
```
Comment (chunweilei) [inline core/src/main/java/org/apache/calcite/rel/core/Aggregate.java:549]: Would there be duplicate values since Set is changed to List?

Comment (liyafan82) [reply core/src/main/java/org/apache/calcite/rel/core/Aggregate.java:549]: @chunweilei Thanks a lot for your feedback.
According to the definition of Rollup (https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/rel/core/Aggregate.java#L508-L513), there can be no duplicate.

**Comment by liyafan82** [comment]
> > > FWIW, I intentionally used `HashMap` rather than `HashSet`. `HashSet` uses a `HashMap` under the covers. I figured I'd save an object allocation. Yours is a bit clearer, a bit less efficient.
> 
> > Thanks a lot for your feedback. It is clever to `HashMap` to avoid object allocation.
> > Fortunately, the value object for `HashSet` is declared static, so it is allocated once, and shared by all `HashSet` objects.
> 
> I think what @julianhyde means is that the HashSet holds a pointer to HashMap, with HashSet object header, it will use 16 bytes more than bare HashMap.

@hsyuan Thanks for your kind reminder. I have reverted the code to use `HashMap`, and made it clear in the comment that this is by design. BTW, why it is 16 more bytes?

**Comment by hsyuan** [comment]
> I have reverted the code to use HashMap, and made it clear in the comment that this is by design. BTW, why it is 16 more bytes?

Suppose in Hotspot 64-bit VM, with COOPS enabled, the object header is 12 bytes, the pointer to HashMap uses 4 bytes. So it is 16 bytes.

**Comment by liyafan82** [comment]
> > I have reverted the code to use HashMap, and made it clear in the comment that this is by design. BTW, why it is 16 more bytes?
> 
> Suppose in Hotspot 64-bit VM, with COOPS enabled, the object header is 12 bytes, the pointer to HashMap uses 4 bytes. So it is 16 bytes.

Awesome! Thanks for your clarification.

**Pull request merged**